Better loading extension and Hot reload

kianzarrin opened the pull request:
see #211 

Currently ~~while~~ when loading another game while in game results in multiple instances of TMPE UI.
I Completed `LoadingExtension.UnLevelUnloading()` To fix this.

Created Hot reload functionality:
 - everything works fine.
 - TMPE traffic rules are reverted to the last time the game was loaded. any extra TMPE rules beyond that are lost (which I think can be a good thing). EDIT: The lane arrows stay the same though. fixing this is blocked by #638
 -  The BaseUI displays assembly version if the TMPE is hot reload.

Test:
 - loaded a game with traffic rules. 
 - Hot-swapped TMPE,
 - Result: cars continue to obey the TMPE rules.
 - added new rules.
 - Result: cars obey the new rules.
 - Hot-swapped TMPE
 - Result: only new rules are gone and cars behave as expected.

EDIT: more tests:
 - loaded a new/saved game
 - quite to main menu
 - loaded a new/saved game again
 -  Result: TMPE is operational (just as it was before my changes)
 - loaded a new/saved game
 - rebuild TMPE for in game hot swap
 - quite to main menu
 - rebuild TMPE again for hot swap in main menu
 - loaded a new/saved game again
 - Result: TMPE operational

![Screenshot (387)](https://user-images.githubusercontent.com/26344691/73615541-66b8d900-4611-11ea-9bb8-f83059e89e21.png)

### reliability:
Note that even in a small mod Hot-swap is not super reliable (AFAIK). but what I have done here is as good as it gets. In all of my tests I did not see hot swap fail.

Note: the game simulation comes to a halt when swapping assemblies which is good for avoiding race conditions :) 

EDIT: i modified `TLM.csproj` the post build events to print something like this:
`THE ASSEMBLY VERSION IS: 1.0.7341.16627 created at  9:14:25.22`

aubergine10 commented on TLM/TLM/UI/MainMenu/VersionLabel.cs:
```diff
@@ -9,6 +10,13 @@ public class VersionLabel : UILabel {
+
+            if(LoadingExtension.Instance.HotReload) {
+                // make it easier to Identify Hot reload.
+                text += " HOT RELOAD " +
```
Would users ever see this? For example, they are subscribed to mod, and have game sitting at Main Menu screen. We push update to mod, Steam downloads it (replacing files on disk with new version).

Or does hotloading only happen if mod files are in `...\Colossal Order\Cities_Skylines\Addons\Mods` folder?

kianzarrin replied:
Changing the assemblies while at main menu does not show hot reload text. only when changing assemblies while in game, hot reload text will appear.

users can see this if they replace their assemblies ... somehow.

`LoadingExtension.Instance.HotReload = InGame()` at `LoadingExtension.OnCreate()` which is only possible while swapping DLLs while in game.

krzychu124 replied:
> users can see this if they replace their assemblies ... somehow.

Steam client won't update anything while in-game, so it can be used by modders - manually/script

kvakvs commented on TLM/TLM/UI/MainMenu/MenuToolModeButton.cs:
```diff
@@ -1,5 +1,6 @@
-﻿namespace TrafficManager.UI.MainMenu {
+namespace TrafficManager.UI.MainMenu {
     using ColossalFramework.UI;
+    using CSUtil.Commons;
```
Added a new import but no other code changes, how does that work?

kianzarrin replied:
i was debugging here and forgot to remove the import.

aubergine10 commented on TLM/TLM/LoadingExtension.cs:
```diff
@@ -333,21 +347,37 @@ public class ManualHarmonyPatch {
+
+                Log.Info("Removing Controls from UI.");
+                if (BaseUI != null) {
+                    BaseUI.Close();
```
Possibly won't matter as it's only devs who are going to use hot reload, but does `BaseUI.Close()` clear all event handlers from any active UI? If not, `Object.Destroy(BaseUI)` is probably going to mem leak. CO's UI stuff doesn't ditch event handlers on destroy.

aubergine10 commented on TLM/TLM/LoadingExtension.cs:
```diff
@@ -333,21 +347,37 @@ public class ManualHarmonyPatch {
-                    Object.Destroy(removeCitizenInstanceButtonExtender, 10f);
+                if (TransportDemandUI != null) {
+                    UIView uiView = UIView.GetAView();
+                    Object.Destroy(TransportDemandUI);
```
Same risk of mem leak here; check if any event handlers are cleared before destroying.

aubergine10 commented on TLM/TLM/UI/MainMenu/VersionLabel.cs:
```diff
@@ -9,6 +10,13 @@ public class VersionLabel : UILabel {
+                // make it easier to Identify Hot reload.
+                text += " HOT RELOAD " +
+                    Assembly.GetExecutingAssembly().GetName().Version;
+            }
```
I'd prefer keeping this; it avoids any doubt

kianzarrin commented:
> Same risk of mem leak here; check if any event handlers are cleared before destroying.

@aubergine10 can you please tell me how an I check/release event handlers?

aubergine10 commented:
> how an I check/release event handlers?

I'm not sure how to do it on CO's UI stuff; but possibly check if they are `null` (if so, no listeners assigned). It would be a PITA to do though. As it's only ever going to be devs doing this, maybe we just accept that there will be some mem leaks? I think having hot loading in `master` branch will save so much time it's worth the rish, but krzychu might disagree.

kianzarrin commented:
> It looks like TrafficManagerTool is leaking, not sure why, but it is not cleared properly from scene (use ModTools -> ToolController to check)
Also noticed delayed removal of CustomPathManager. It works but looks wierd.
Steps to reproduce:
    Load save game
    perform hot reload.
    after few seconds you should see: CustomPathManager: OnDestroy with old log time almost at the end of TMPE.log

I know about CustomPathManager . The delayed removal is intentional. The code tries to remove original path manager on load. so it removes `CustomPathManger` on reload. I got null reference error when I tried to remove `CustomPathmanager` on unload because the loading extension of the new mod expects there to be `CustomPathManager`. So i decided not to remove it. I guess the game remove the path manager when quieting to main menu? so I am not sure if there is going to ever be a memleak.

I checkout `TrafficManagerTool`

kianzarrin commented:
@krzychu124 TrafficManagerTool is automatically created by UIBase on access. but it is not automatically destroyed. so I manually destroyed it. it should be fixed now.  I changed an unnecessary static field in TrafficManagerTool  to instance field (`_activeSubTool`). 

I changed unload order a bit to be reverse of load order. 

I tested my changes by hot-swapping traffic manager in game and in the main menu. i load/reload game several times. no exceptions no errors.

kianzarrin commented on TLM/TLM/LoadingExtension.cs:
```diff
@@ -16,10 +16,30 @@ namespace TrafficManager {
+
+        internal bool InGameHotReload { get; private set; } = false;
+
+        internal static AppMode currentMode => SimulationManager.instance.m_ManagersWrapper.loading.currentMode;
```
@krzychu124 SceneManager.GetActiveScene.name is "Game" even at the stage when LoadingExtension.OnCreated() when loading game from main menu. 

currentMode == AppMode.Game is not game while loading a game from main menu so it is useful for me. 

I was looking for  some code for my `InGame()` function that is false unless everything is loaded and game has started. Currently I need to take a note  of `currentMode` in `LoadingExtenson.OnCreated()` before it changes to something not useful.

kianzarrin commented on TLM/TLM/LoadingExtension.cs:
```diff
@@ -359,25 +355,30 @@ public class ManualHarmonyPatch {
                 Log.Info("Removing Controls from UI.");
                 if (BaseUI != null) {
                     BaseUI.Close(); // Hide the UI ASAP
-                    Object.Destroy(BaseUI);
```
why I don't see traffic manager in modtools?
![Screenshot (424)](https://user-images.githubusercontent.com/26344691/74087891-04405c80-4a99-11ea-8f97-c8b24473c0f0.png)

Sorry about the sloppy work. I got totally confused! you are spot on. I shall release trafficManger from UIBase destructor.

kianzarrin replied:
OK this is fixed :) 
![Screenshot (425)](https://user-images.githubusercontent.com/26344691/74088039-a0b72e80-4a9a-11ea-8c2b-5d3eee1df0d0.png)
```
Debug 52.4193354: UIBase destructor is called.
Debug 52.4235121: LoadingExtension.DisableTool: called
Info 52.4301887: Removing Traffic Manager Tool.
Info 58.5976456: TrafficManagerTool.OnDestroy() called
Info 58.8597243: TM:PE disabled.
```


I cleaned up the UIBase destruction path a little bit. That and your suggestion fixed the problem. I think part of the problem was traffic manager destructor was never called because of some out of place null reference checks.

kianzarrin commented:
Wow I have write access to merge it. should I?

aubergine10 commented:
Yup go ahead and merge. :) 

Also you can now create branches on the main repo if desired (`master` branch is locked from direct merging; merges always go through PR review).
